### PR TITLE
Can bump open firelocks with crowbar in active hand

### PR DESCRIPTION
**Pull request by cacogen**
## About The Pull Request

https://user-images.githubusercontent.com/25089914/155260045-89aad25e-5b09-4889-9ad4-f28c49598814.mp4

If you bump into a firelock with a crowbar in your active hand (or as your active module as a cyborg), it will open as though you'd clicked on it.

~~Got most of this code from automatically mining bumped turfs.~~ I made it a component, although I'm not sure if it's any good or not.

## Why It's Good For The Game

Makes navigating firelock hell slightly quicker, without repeated clicking. They still stop you for a second if you have a crowbar and GBJ you if you don't.

## Changelog

:cl: cacogen
qol: Firelocks will open when walked into with a crowbar or fireaxe in your active hand
qol: Resonator can be bump mined with (way less unwieldy than spam-clicking with the thing)
/:cl:

### CODE REVIEW

**Comment by Razharas** [comment]
make this generic for any item on any bump target

**Comment by cacogen** [comment]
I think that would lead to some undesirable interactions, like bumping into an airlock with a welder starting to fix it or weld it closed. 

I think it only makes sense in certain situations, like when you know the player wants to move onto the same turf as the atom after using a tool on it. And it's only really needed when it's likely they'll be doing it repeatedly (e.g. mining turfs, opening firelocks). Another case I can think of is opening depowered airlocks. 

I don't know how maintainers like oranges and LemonInTheDark are going to feel about this, though. I don't think having to click repeatedly is a desirable or necessary part of game balance but they may disagree. And it's made worse by lag. The click cooldown causes like one click to go through each second, and some of those are dropped due to lag. Movement is more responsive.

**Comment by GuillaumePrata** [comment]
Bumping into an airlock will hold it open then?
Because if it is just the quick open that let it auto drops after a second, being smashed by firelocks again will get old real fast.

**Comment by cacogen** [comment]
It just acts like a left-click. So it holds it open, yeah.

**Comment by SuperNovaa41** [comment]
> I think that would lead to some undesirable interactions, like bumping into an airlock with a welder starting to fix it or weld it closed.
> 
> I think it only makes sense in certain situations, like when you know the player wants to move onto the same turf as the atom after using a tool on it. And it's only really needed when it's likely they'll be doing it repeatedly (e.g. mining turfs, opening firelocks). Another case I can think of is opening depowered airlocks.
> 
> I don't know how maintainers like oranges and LemonInTheDark are going to feel about this, though. I don't think having to click repeatedly is a desirable or necessary part of game balance but they may disagree. And it's made worse by lag. The click cooldown causes like one click to go through each second, and some of those are dropped due to lag. Movement is more responsive.

I believe he means turn it into an element, so it can be applied to any tool. so instead of copy pasting this code, you would just apply the element to the crowbar, with a callback to pry whenever the atom is a firelock (or something to that extent)

**Comment by DrDiasyl** [comment]
Rebalance? This sounds like more of quality of life tho.

**Comment by moocowswag** [comment]
Ya its definitely a QoL, its not like its letting you bypass click cooldowns or anything, infact I dont think clicking on a closed firelock even gives you a cooldown and youre able to just quickly spam click all of the ones near you open.

**Comment by cacogen** [comment]
I hope this gets reviewed now

**Comment by timothymtorres** [comment]
This is a nice QoL change. Sometimes people are too slow to click on a firelock and move through it in time.  Whoever reviews this PR needs to remove the Balance label.

**Comment by TheBonded** [comment]
this is going to save so much time and annoyances

**Comment by cacogen** [comment]
I don't think it's ever going to be merged though...

**Comment by TheBonded** [comment]
> I don't think it's ever going to be merged though...

ping the maintainers

File: code/game/machinery/doors/firedoor.dm [hunk 1]
```diff
@@ -290,11 +294,12 @@
 	obj_flags |= EMAGGED
 	INVOKE_ASYNC(src, .proc/open)
 
-/obj/machinery/door/firedoor/Bumped(atom/movable/AM)
+/obj/machinery/door/firedoor/Bumped(atom/movable/movable)
 	if(panel_open || operating)
 		return
 	if(!density)
 		return ..()
+	SEND_SIGNAL(src, COMSIG_ATOM_BUMPED, movable)
```
Comment (LemonInTheDark) [inline code/game/machinery/doors/firedoor.dm:302]: this is really horrible you know

**Comment by LemonInTheDark** [comment]
> Makes navigating firelock hell slightly quicker, without repeated clicking. They still stop you for a second if you have a crowbar and GBJ you if you don't.

do we really need to optimize away one click?

File: code/game/machinery/doors/firedoor.dm (hunk 1 above)
Comment (cacogen) [reply code/game/machinery/doors/firedoor.dm:302]: I don't understand the significance of it, to tell you the truth. It doesn't return parent by default, so how are you supposed to get all the way up the chain to where it calls BUMPED at the atom level? Do I have to make it call parent and then fix whatever unintended functionality arises from that?

**Comment by cacogen** [comment]
It's like you want me to add stat tracking for the number of firelocks crowbarred open by each player each round so we can see if on average it's a significant amount or not

File: code/game/machinery/doors/firedoor.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/game/machinery/doors/firedoor.dm:302]: image

Comment (cacogen) [reply code/game/machinery/doors/firedoor.dm:302]: I've gone through that and from what I can tell by calling parent the only issue that arises with firelocks is throwing your ID to open them which I've fixed in the code I'm about to test then push.

**Comment by LemonInTheDark** [comment]
I don't need like, empirical data. I can imagine opening like 20 firelocks might be a bit annoying, but not everything that's annoying is bad. 
Fire alarms have had a ton of the edges sanded off recently, and I don't want to just make them not a problem, or not at least a bit frustrating. I don't like that pattern.

**Comment by optimumtact** [comment]
there does not appear to be any maintainer wanting to merge this.

**Pull request closed without merging**